File: Python/SortingAlgorithms/merge_sort.py
from typing import TypeVar, List

T = TypeVar("T")

def merge_sort(arr: List[T]) -> None:
    merge_sort_recursive(arr, 0, len(arr) - 1)
    

def merge_sort_recursive(arr: List[T], p: int, r: int) -> None:
    if p >= r: return
    q = p + (r - p) // 2
    merge_sort_recursive(arr, p, q)
    merge_sort_recursive(arr, q + 1, r)
    merge(arr, p, q, r)



def merge(arr: List[T], p: int, q: int, r: int) -> None:
    if arr[q] < arr[q+1]: return

    nL = q - p + 1
    nR = r - q

    L = arr[p : p + nL]
    R = arr[q + 1 : q + 1 + nR]

    i, j, k = 0, 0, p

    while i < nL and j < nR:
        if L[i] < R[j]:
            arr[k] = L[i]
            i = i + 1
        else:
            arr[k] = R[j]
            j += 1
        k += 1
       
    while i < nL:
        arr[k] = L[i]
        i += 1
        k += 1

    while j < nR:
        arr[k] =R[j]
        j += 1
        k += 1

File: C++/SortingAlgorithms/MergeSort.cpp
#include "MergeSort.h"


template<typename T>
void merge_sort(std::vector<T>& arr)
{
	merge_sort_recursive(arr, 0, arr.size() - 1);
}

template<typename T>
void merge_sort_recursive(std::vector<T>& arr, int p, int r)
{
	if (p >= r) return;

	int q = p + (r - p) / 2;

	merge_sort_recursive(arr, p, q);
	merge_sort_recursive(arr, q + 1, r);

	merge(arr, p, q, r);
}

template<typename T>
void merge(std::vector<T>& arr, int p, int q, int r)
{
	int nL = q - p + 1;
	int nR = r - q;

	auto startL = arr.begin() + p;
	auto endL = startL + nL;

	std::vector<T> L(startL, endL);

	auto startR = arr.begin() + q + 1;
	auto endR = startR + nR;

	std::vector<T> R(startR, endR);

	int i = 0, j = 0, k = p;

	while (i < nL && j < nR) {

		if (L[i] < R[j]) {
			arr[k] = L[i];
			i++;
		}
		else {
			arr[k] = R[j];
			j++;
		}
		k++;
	}

	while (i < nL) {
		arr[k] = L[i];
		i++;
		k++;
	}

	while (j < nR) {
		arr[k] = R[j];
		j++;
		k++;
	}
}

template void merge_sort(std::vector<int>&);
template void merge_sort_recursive(std::vector<int>&, int, int);
template void merge(std::vector<int>&, int, int, int);